Issue #471 Route `POST /units-for-slot`

elpiel opened the pull request:
Part of #471 

The initial move of the `GET /units-for-slot` route from the Supermarket to sentry `POST /units-for-slot`

Notable changes to primitives:
- EventType is used for targeting input
- IPFS is used for targeting input instead of string
- Impression & Click Events now have mandatory ad_unit & ad_slot instead of Option
- Campaign's Pricing Bounds now drops the Option and uses a Hashmap instead of struct with separate fields for each EventType

elpiel committed:
sentry - units-for-slot - initial move from supermarket

elpiel committed:
sentry - platform - PlatformApi
- Move from sentry - routes - units-for-slot
- Application - add PlatformApi
- Fix all calls of Application::new()

elpiel committed:
rustfmt & clippy

elpiel committed:
sentry - routers - some docs

elpiel committed:
sentry - Cargo - postgres-types - add array-impls feature

elpiel committed:
sentry - routes - fix method of channel last-approved

elpiel committed:
sentry - PlatformApi - remove logger

elpiel committed:
primitives - Config - limits & platform config
- Config - limits for units-for-slot
- Config - deserializae milliseconds to std duration
- dev.toml / ganache.toml - add values for limits & platform config

elpiel committed:
sentry - Platform - clean up impl
- primitives - platform - AdUnit responses

elpiel committed:
- primitives - AdSlot uses IPFS struct
- primitives - sentry - impl From<EventType> for String

elpiel committed:
sentry - db - campaign:
- units_for_slot_get_campaigns
- CampaignRemaining::get_multiple_with_ids

elpiel committed:
sentry - routes - units-for-slot - main V5 logic is complete

elpiel committed:
Merge branch 'aip-61-adex-v5' into issue-471-route-units-for-slot

simzzz commented on sentry/src/main.rs:
```diff
@@ -106,6 +107,14 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
 
     let campaign_remaining = CampaignRemaining::new(redis.clone());
 
+    // todo: Make platform_url configurable! Load from config or pass with env. variable
```
We have those values in the config. Can't we just load them from config now?

elpiel replied:
it's fixed now.

simzzz commented on primitives/src/config.rs:
```diff
@@ -37,7 +38,7 @@ impl Default for Environment {
 
 #[derive(Serialize, Deserialize, Debug, Clone)]
-#[serde(rename_all(serialize = "SCREAMING_SNAKE_CASE"))]
+// TODO: use `milliseconds_to_std_duration` for all u32 values with "In milliseconds" in docs.
```
Is the `rename_all` attribute removed on purpose?

elpiel replied:
yep, it is. I've decided to drop it since we are no longer bound to the flat structure of the Config

simzzz commented on sentry/src/db/campaign.rs:
```diff
@@ -210,8 +216,51 @@ pub async fn update_campaign(pool: &DbPool, campaign: &Campaign) -> Result<Campa
+    params.push(Box::new(creator));
+
+    // Campaign.active.to
+    where_clauses.push("active_to >= $4".into());
```
`active_to` is added twice

elpiel replied:
fixed.

simzzz commented on sentry/src/routes/units_for_slot.rs:
```diff
@@ -0,0 +1,422 @@
+    Application, ResponseError,
+};
+
+pub(crate) static CLOUDFLARE_IPCOUNTY_HEADER: Lazy<HeaderName> =
```
The variable is `IPCOUNTY`, should be `IPCOUNTRY`

elpiel replied:
fixed

elpiel committed:
sentry - create PlatforApi from config values

elpiel committed:
sentry - Fix PR comments:
- sentry - db -campaign - duplicate where cluase
 sentry - routes - units for slot - typo in static name

elpiel committed:
primitive improvements:
- primitives - campaign - Pricing Bounds
- primitives - sentry - Event - ad_unit and ad_slot mandatory
- primitives - targeting - input - use IPFS & EventType

elpiel committed:
fixes, formatting and clippy around:
- changes on targeting input
- Event ad_unit & ad_slot
- analytics

elpiel committed:
sentry - instanciate PlatformApi using Config

elpiel commented on primitives/src/campaign.rs:
```diff
@@ -217,34 +216,36 @@ impl Campaign {
-        self.pricing_bounds
-            .as_ref()
-            .and_then(|bound| bound.get(event))
+    pub fn pricing(&self, event: EventType) -> Option<&Pricing> {
+        self.pricing_bounds.get(&event)
     }
 }
```
This is redundant, so it should be remove soon

elpiel commented on primitives/src/campaign.rs:
```diff
@@ -253,36 +254,33 @@ mod pricing {
-    }
+    #[derive(Default, Debug, PartialEq, Eq, Serialize, Deserialize, Clone)]
+    #[serde(transparent)]
+    pub struct PricingBounds(HashMap<EventType, Pricing>);
```
Pricing bounds is now using the EventType

elpiel committed:
primitives - supermarket - remove old V4 structs

simzzz commented on test_harness/src/lib.rs:
```diff
@@ -977,15 +993,25 @@ pub mod run {
+        let logger = new_logger(&validator.sentry_logger_prefix);
         let campaign_remaining = CampaignRemaining::new(redis.clone());
 
+        // todo: Make platform_url configurable! Load from config or pass with env. variable
+        let platform_url = "https://todo-local-platform"
+            .parse()
+            .expect("Bad ApiUrl, load from Config?");
+        // todo: Make keep_alive_interval configurable!
+        let platform_api = PlatformApi::new(platform_url, std::time::Duration::from_secs(3))
+            .expect("Should make PlatformApi");
```
```suggestion
       let platform_api = PlatformApi::new(
          config.platform.url.clone(),
          config.platform.keep_alive_interval,
      )
      .expect("Failed to build PlatformApi");
```

We can use the values from config here as well.

simzzz reviewed (changes requested):
One more thing I forgot to comment during my previous review. Other than that LGTM.

elpiel committed:
test_harness - use config keep alive for PlatformApi

Co-authored-by: Simeon Nakov <simeon@ambire.com>

elpiel committed:
test_harness - fix platfrom config
